### PR TITLE
Check for falling `float` nodes in liquid transform

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
Fixes https://github.com/minetest/minetest/issues/12839. There should be minimal effect on liquid performance, but I haven't actually benchmarked.

## To do

This PR is Ready for Review.

## How to test

Make sure that liquids still work and that `testnodes:falling_float` falls when the liquid under it disappears.

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -610,7 +612,9 @@ void ServerMap::transformLiquids(std::map<v3s16, MapBlock*> &modified_blocks,
 			v3s16 npos = p0 + liquid_6dirs[i];
 			NodeNeighbor nb(getNode(npos), nt, npos);
 			const ContentFeatures &cfnb = m_nodedef->get(nb.n);
-			switch (m_nodedef->get(nb.n.getContent()).liquid_type) {
+			if (nt == NEIGHBOR_UPPER && cfnb.floats)
```
Comment (SmallJoker) [inline src/map.cpp:615]: This should be run for all falling nodes which could be generated by the mapgen or manually with worldedit or other mods.

Comment (TurkeyMcMac) [reply src/map.cpp:615]: In my view that would break compatibility.

Comment (SmallJoker) [reply src/map.cpp:615]: Compatibility with what?

Comment (TurkeyMcMac) [reply src/map.cpp:615]: Say you explore some caves with dynamic_liquid enabled. Due to liquid movement, suddenly the suspended sand in proximity to the water would fall without you having to punch it.

Comment (sfan5) [reply src/map.cpp:615]: I think this would be a bad idea too, but in any case it's for another PR

**Review by SmallJoker — APPROVED** [review]
Works.

**Comment by sfan5** [comment]
A bit unfortunate that we have to hardcode engine support for something that has so far been Lua-side only but I don't see a better way.

**Comment by TurkeyMcMac** [comment]
The review should be redone with the new commit.

**Comment by SmallJoker** [comment]
Wouldn't it be more efficient if you set a boolean inside the loop, say `have_floating_attached_above = true` inside this loop:

https://github.com/minetest/minetest/blob/dc4f8ffc9e316bdb241e8a4820f4ef26d0ac71d9/src/map.cpp#L615-L616

And run the callback next to `on_flood`, but when the current node turns to CONTENT_AIR ?

https://github.com/minetest/minetest/blob/dc4f8ffc9e316bdb241e8a4820f4ef26d0ac71d9/src/map.cpp#L760-L780

**Comment by TurkeyMcMac** [comment]
I'm not sure if the falling checks should be done in the main loop, since the checks will change nodes at a time when lighting has not yet been updated. Could this cause lighting errors?

**Comment by SmallJoker** [comment]
Well then, add it to the vector in that location because the content ID change is guaranteed there.

**Review by SmallJoker — APPROVED** [review]
Works on my machine.

**Pull request merged**